index/scorch/merge.go

```diff
@@ -129,8 +146,65 @@ func (s *Scorch) parseMergePlannerOptions() (*mergeplan.MergePlanOptions,
 	return &mergePlannerOptions, nil
 }
 
+type mergerKick struct {
+	moptions *mergeplan.MergePlanOptions
+	doneCh   chan struct{}
+}
+
+func (s *Scorch) RequestMerge(mo *mergeplan.MergePlanOptions) error {
+	// check whether force merge is already under processing
+	s.rootLock.Lock()
+	if atomic.LoadUint64(&s.stats.TotFileMergeForceOpsStarted) >
+		atomic.LoadUint64(&s.stats.TotFileMergeForceOpsCompleted) {
+		s.rootLock.Unlock()
+		return fmt.Errorf("compaction already in progress")
+	}
+	s.rootLock.Unlock()
```

Comment:
now that time has passed since looking at this PR... i this is going to need a little bit more thought...

when i see lots of atomics being used, along with while they're embedded inside of locking... my gut feel worry instincts start tingling :-)

aside: it's slightly amazing that you all the way in Bangalore can cause this time-and-space delayed effect to happen halfway around the world!  we live in amazing times! :-)

quick thought is that reading these atomics while under a lock might not be helpful... especially if the mutations of these atomics are not under a lock.

And, then, if the next proposed move might be perhaps to refactor these atomics into the protection of a lock, then the thought would be... why would we need atomics?

anyways, stuff like these appear in my mind

...as

Reply:
yes..there is a raciness issue there.. thanks @steveyen .
I need to add the exit on whether "force merge is already under processing" check inside the go-routine at line no:164.5 as well as there would be a random delay in the actual scheduling of a go-routine at runtime. 
So with the current code  - there could be N - concurrent go-routines all passing the original checks there at line:157 and triggering a forced merge routine..
Now it looks like, I don't really need a s.rootLock at all, all atomic checks will be enough...is what comes to mind.. With the additional check inside the routine at line:164.5, only one routine will win the force merge..